integration/erc20contract/ObsERC20.sol

```diff
@@ -14,12 +16,37 @@ contract ObsERC20 is ERC20 {
     }
 
     function balanceOf(address account) public view virtual override returns (uint256) {
-        // Respond to balance requests made by the account owner only.
-        // In case the requester spoofs the "from" of the call, they will not be able to read
-        // the result since it will be returned encrypted with the viewing key of the declared "from".
-        // todo - figure out the correct condition to work with other contracts
-//        require(msg.sender == account);
-        return super.balanceOf(account);
+        // 1. Human owner of an account asking for the balance.
+        // 2. Human owner of an account interacting with a smart contract which in turn asks for the balance of the original asker.
+        // Note: In case the requester spoofs the "from" of the call, they will not be able to read
+        //  the result since it will be returned encrypted with the viewing key of the declared "from".
+        if (tx.origin == account) {
+            return super.balanceOf(account);
+        }
+
+        // 3. Contract asking for its own balance.
+        if (msg.sender == account) {
```

Comment:
Would be clearer IMO if we combined these checks into one check, if that's possible?

Reply:
it felt a bit more explicit if they're separate, with individual comments.
We'll have to think about gas cost as well.